packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -794,10 +602,240 @@ KernelAsset _targetLocationSingleArchitecture(
   );
 }
 
-/// Extract the [Target] from a [TargetPlatform].
+Map<NativeCodeAssetImpl, KernelAsset> _assetTargetLocationsForOS(OS targetOS,
+    List<NativeCodeAssetImpl> codeAssets, bool flutterTester, Uri buildUri) {
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      final Uri? absolutePath = flutterTester ? buildUri : null;
+      return _assetTargetLocationsWindowsLinux(codeAssets, absolutePath);
+    case OS.macOS:
+      final Uri? absolutePath = flutterTester ? buildUri : null;
+      return assetTargetLocationsMacOS(codeAssets, absolutePath);
+    case OS.iOS:
+      return assetTargetLocationsIOS(codeAssets);
+    case OS.android:
+      return assetTargetLocationsAndroid(codeAssets);
+    default:
+      throw StateError('This should be unreachable.');
+  }
+}
+
+Future<void> _copyNativeCodeAssetsForOS(
+    OS targetOS,
+    Uri buildUri,
+    build_info.BuildMode buildMode,
+    FileSystem fileSystem,
+    Map<NativeCodeAssetImpl, KernelAsset> assetTargetLocations,
+    String? codesignIdentity,
+    bool flutterTester) async {
+  final List<NativeCodeAssetImpl> codeAssets =
+      assetTargetLocations.keys.toList();
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      assert(codesignIdentity == null);
+      await _copyNativeCodeAssetsToBundleOnWindowsLinux(
+        buildUri,
+        assetTargetLocations,
+        buildMode,
+        fileSystem,
+      );
+    case OS.macOS:
+      if (flutterTester) {
+        await copyNativeCodeAssetsMacOSFlutterTester(
+          buildUri,
+          fatAssetTargetLocationsMacOS(codeAssets, buildUri),
+          codesignIdentity,
+          buildMode,
+          fileSystem,
+        );
+      } else {
+        await copyNativeCodeAssetsMacOS(
+          buildUri,
+          fatAssetTargetLocationsMacOS(codeAssets, null),
+          codesignIdentity,
+          buildMode,
+          fileSystem,
+        );
+      }
+    case OS.iOS:
+      await copyNativeCodeAssetsIOS(
+        buildUri,
+        fatAssetTargetLocationsIOS(codeAssets),
+        codesignIdentity,
+        buildMode,
+        fileSystem,
+      );
+    case OS.android:
+      assert(codesignIdentity == null);
+      await copyNativeCodeAssetsAndroid(
+          buildUri, assetTargetLocations, fileSystem);
+    default:
```

Comment:
This isn't defined in flutter_tools, but is there any reason we aren't just using enhanced enums for `OS` and `OSImpl`? This would allow for us to perform a static exhaustiveness check for all these switches without requiring us to manually raise a `StateError` everywhere.

Reply:
Fair enough. Should we maybe be throwing `UnimplementedError` instead then?

Reply:
What Daco says. Adding more entries to the enum shifts from being a compile-time error to being a runtime error (and only if the new entry is used)

Reply:
Done.